pkg/appgateserver/server.go

```diff
@@ -175,75 +175,88 @@ func (app *appGateServer) ServeHTTP(writer http.ResponseWriter, request *http.Re
+	rpcType := rpcdetect.GetRPCType(poktHTTPRequest)
 
-	// Get the type of the request by doing a partial unmarshal of the payload
-	// TODO_BLOCKER(#511): Move request RPCType detection to shannon-sdk
-	requestType, err := partials.GetRequestType(ctx, requestPayloadBz)
-	if err != nil {
-		app.replyWithError(ctx, requestPayloadBz, writer, serviceId, "unknown", ErrAppGateHandleRelay)
-		// TODO_IMPROVE: log additional info?
-		app.logger.Error().Err(err).Msg("failed getting request type")
+	// Add newly available fields to the error logger.
+	errorLogger = errorLogger.
+		Str("rpc_type", rpcType.String()).
+		Str("payload", string(poktHTTPRequest.BodyBz))
 
-		return
-	}
+	app.logger.Debug().
+		Str("service_id", serviceId).
+		Str("rpc_type", rpcType.String()).
+		Msg("detected rpc type")
 
 	// Determine the application address.
 	appAddress := app.signingInformation.AppAddress
 	if appAddress == "" {
 		appAddress = request.URL.Query().Get("applicationAddr")
 	}
 	if appAddress == "" {
-		app.replyWithError(ctx, requestPayloadBz, writer, serviceId, requestType.String(), ErrAppGateMissingAppAddress)
-		// TODO_IMPROVE: log additional info?
-		app.logger.Error().Msg("no application address provided")
+		// If no application address is provided, reply with an error response.
+		app.replyWithError(ErrAppGateMissingAppAddress, poktHTTPRequest, serviceId, rpcType, writer)
+		errorLogger.Err(ErrAppGateMissingAppAddress).Msg("no application address provided")
 
 		return
 	}
 
-	// Put the request body bytes back into the request body.
-	request.Body = io.NopCloser(bytes.NewBuffer(requestPayloadBz))
+	// Create a requestInfo struct to pass to the handleSynchronousRelay method.
+	reqInfo := &ruquestInfo{
+		appAddress:  appAddress,
+		serviceId:   serviceId,
+		rpcType:     rpcType,
+		poktRequest: poktHTTPRequest,
+		requestBz:   requestBz,
+	}
 
-	// TODO_IMPROVE(@red-0ne): Add support for asynchronous relays, and switch on
+	// TODO_IMPROVE(@red-0ne, #40): Add support for asynchronous relays, and switch on
 	// the request type here.
-	// TODO_RESEARCH: Should this be started in a goroutine, to allow for
-	// concurrent requests from numerous applications?
-	if err := app.handleSynchronousRelay(
-		ctx, appAddress, serviceId, requestType, request, writer); err != nil {
-
+	if err := app.handleSynchronousRelay(ctx, reqInfo, writer); err != nil {
 		// Reply with an error response if there was an error handling the relay.
-		app.replyWithError(ctx, requestPayloadBz, writer, serviceId, requestType.String(), err)
-		// TODO_IMPROVE: log additional info?
-		app.logger.Error().Err(err).Msg("failed handling relay")
+		app.replyWithError(err, poktHTTPRequest, serviceId, rpcType, writer)
+		errorLogger.Err(err).
+			Str("application_addr", appAddress).
+			Msg("failed handling relay")
 
 		return
 	}
 
-	// TODO_IMPROVE: log additional info?
-	app.logger.Info().Msg("request serviced successfully")
+	app.logger.Info().
+		Str("service_id", serviceId).
```

Comment:
DItto. I don't think we need to add `service_id` and you can make `rpc_type` a recurring field of the local logger.